### PR TITLE
Extract token count from streaming Fireworks requests

### PR DESCRIPTION
**Pull request by rafax**
Extract token counts from streaming Fireworks requests - [context](https://sourcegraph.slack.com/archives/C060RPB9H1T/p1704283962290259?thread_ts=1704213101.928429&cid=C060RPB9H1T). 
Non-streaming requests already set token counts correctly.

Closes https://github.com/sourcegraph/sourcegraph/issues/56704

## Test plan

- tested locally
- unit tests added

### CODE REVIEW

**Review by cbart — APPROVED** [review]

**Review by chwarwick — APPROVED** [review]

File: cmd/cody-gateway/internal/httpapi/completions/fireworks.go [hunk 1]
```diff
@@ -211,10 +212,20 @@ func (f *FireworksHandlerMethods) parseResponseAndUsage(logger log.Logger, reqBo
 		if len(event.Choices) > 0 {
 			completionUsage.characters += len(event.Choices[0].Text)
 		}
+		// these are only included in the last message, so we're not worried about overwriting
```
Comment (unknwon) [inline cmd/cody-gateway/internal/httpapi/completions/fireworks.go:215]: nit

```suggestion
		// These are only included in the last message, so we're not worried about overwriting
```

😛

**Review by unknwon — APPROVED** [review]

File: cmd/cody-gateway/internal/httpapi/completions/fireworks.go [hunk 2]
```diff
@@ -211,10 +212,20 @@ func (f *FireworksHandlerMethods) parseResponseAndUsage(logger log.Logger, reqBo
...
+			promptUsage.tokens = event.Usage.PromptTokens
+		}
+		if event.Usage.CompletionTokens > 0 {
+			completionUsage.tokens = event.Usage.CompletionTokens
+		}
 	}
 	if err := dec.Err(); err != nil {
 		logger.Error("failed to decode Fireworks streaming response", log.Error(err))
 	}
+	if completionUsage.tokens <= 0 || promptUsage.tokens <= 0 {
```
Comment (bobheadxi) [inline cmd/cody-gateway/internal/httpapi/completions/fireworks.go:226]: We set these to -1 as sentinel values in L193

```suggestion
	if completionUsage.tokens == -1 || promptUsage.tokens == -1 {
```

**Review by bobheadxi — APPROVED** [review]
also edited PR description to link to existing issue :)

File: cmd/cody-gateway/internal/httpapi/completions/fireworks.go (hunk 2 above)
Comment (rafax) [reply cmd/cody-gateway/internal/httpapi/completions/fireworks.go:226]: Fair, I initially thought we can override the `-1` with zeros, but the way the code is written now it's either -1 or >0

**Pull request merged**